src/security/KeyData.cc

```diff
@@ -42,6 +43,40 @@ Security::KeyData::loadX509CertFromFile()
         return false;
     }
 
+    try {
+        // selected bundled certificates in sending order: wireCerts = cert + chain
+        CertList wireCerts(1, cert);
+
+        while (const auto bundledCert = Ssl::ReadOptionalCertificate(bio)) {
+            assert(!wireCerts.empty()); // this->cert is there (at least)
+
+            // We cannot chain any certificate after a self-signed certificate. This
+            // check also protects the IssuedBy() check below from adding duplicated
+            // (i.e. listed multiple times in the bundle) self-signed certificates.
+            if (SelfSigned(*wireCerts.back())) {
```

Comment:
Note that we are still sending self-signed certificates. No changes in that area (yet). We are also still avoiding sending the same self-signed certificate twice. The code got simpler because we no longer have to deal with loading the first certificate twice.

Reply:
> Note that we are still sending self-signed certificates. No changes in that area (yet).

That specific change is now proposed in a dedicated PR: #1173.